File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/activity/performance_type_controller.py
import inspect
import logging
from typing import Union
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.shortcuts import render, redirect, get_object_or_404
from django.http import HttpRequest, HttpResponse
from .._traits.controller import Controller
from ....Models.activity.performance_type import PerformanceType
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception

logger = logging.getLogger(__name__)

class PerformanceTypeController(Controller):
  
  @classmethod
  def _set_auth(cls, request: HttpRequest, perm:str) -> Union[bool, Exception]:
    self = cls()
    self.request = request
    self.authorize(f'{perm} performance type')
    return True

  @classmethod
  def index(cls, request: HttpRequest) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      PerformanceTypeController._set_auth(request, 'manage')
      if request.user.type != 'company':
        messages.error(request,'Permission denied.') 
        return redirect(get_redirect_url(request))
      types = PerformanceType.objects.filter(created_by=request.user.creator_id())
      return render(request,'performanceType/index.html',{'types':types})
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def create(cls, request: HttpRequest) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      PerformanceTypeController._set_auth(request, 'create')
      return render(request,'performanceType/create.html')
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def store(cls, request: HttpRequest) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      PerformanceTypeController._set_auth(request, 'create')
      if request.user.type != 'company':
        messages.error(request,'Permission denied.') 
        return redirect(get_redirect_url(request))
      name = request.POST.get('name')
      if not name:
        messages.error(request,'Name is required.') 
        return redirect(get_redirect_url(request))
      PerformanceType.objects.create(
        name=name,created_by=request.user.creator_id()
      )
      messages.success(request,'Performance Type successfully created.')
      return redirect('performance_type_index')
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def show(cls, request: HttpRequest, performance_type_id) -> HttpResponse:
    # no explicit show view redirect back
    return redirect(get_redirect_url(request))

  @classmethod
  def edit(cls, request: HttpRequest, performance_type_id) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      PerformanceTypeController._set_auth(request, 'edit')
      performance_type = get_object_or_404(PerformanceType,pk=performance_type_id)
      return render(request,'performanceType/edit.html',{'performanceType':performance_type})
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def update(cls, request: HttpRequest, performance_type_id) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      PerformanceTypeController._set_auth(request, 'edit')
      performance_type = get_object_or_404(PerformanceType,pk=performance_type_id)
      if request.user.type != 'company':
        messages.error(request,'Permission denied.') 
        return redirect(get_redirect_url(request))
      name = request.POST.get('name')
      if not name:
        messages.error(request,'Name is required.') 
        return redirect(get_redirect_url(request))
      performance_type.name = name
      performance_type.created_by = request.user.creator_id()
      performance_type.save()
      messages.success(request,'Performance Type successfully updated.')
      return redirect('performance_type_index')
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

  @classmethod
  def destroy(cls, request: HttpRequest, performance_type_id) -> HttpResponse:
    REF = f'{cls.__name__}::{inspect.currentframe().f_code.co_name}'
    try:
      PerformanceTypeController._set_auth(request, 'delete')
      performance_type = get_object_or_404(PerformanceType,pk=performance_type_id)
      if request.user.type != 'company':
        messages.error(request,'Permission denied.') 
        return redirect(get_redirect_url(request))
      performance_type.delete()
      messages.success(request,'Performance Type successfully deleted.')
      return redirect('performance_type_index')
    except PermissionDenied as err:
      return default_permission_denial(request,err,ref=REF)
    except Exception as err:
      return default_undefined_exception(request,err,ref=REF)

File: _inc/laravel/app/Http/Controllers/Activity/PerformanceTypeController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
  DatabaseConstants,
  PermissionsConstants,
  UsersConstants,
  ViewsConstants
};
use App\Models\PerformanceType;
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Facades\{Log, Route, Validator, View as ViewFacade};
use Illuminate\View\View;

class PerformanceTypeController extends Controller
{
  private const PERM_CREATE = PermissionsConstants::CRT_PRF_TP;
  private const PERM_DELETE = PermissionsConstants::DEL_PRF_TP;
  private const PERM_EDIT = PermissionsConstants::ED_PRF_TP;
  private const PERM_MANAGE = PermissionsConstants::MNG_PRF_TP;

  public function index(Request $request): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::PFM_TP . '.' . $action;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizeCompany($req, self::PERM_MANAGE)) return redirect()->back();
      try {
        $creatorId = $req->user()->creatorId();
        $fetchStart = microtime(true);
        $types = PerformanceType::where(DatabaseConstants::COL_TABLE_CREATOR, $creatorId)->get();
        $this->logExecutionTime($fetchStart, $action, 'fetchPerformanceTypes');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        Log::info("[{$class}::{$action}] complete", ['count' => is_countable($types) ? count($types) : null]);
        return view($viewPath, compact('types'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function create(Request $request): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::PFM_TP . '.' . $action;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizeCompany($req, self::PERM_CREATE)) return redirect()->back();
      try {
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        $renderStart = microtime(true);
        $resp = view($viewPath);
        $this->logExecutionTime($renderStart, $action, 'renderPerformanceTypeCreate');
        Log::info("[{$class}::{$action}] complete");
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function store(Request $request): RedirectResponse|JsonResponse|null
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['creator_id' => $req->user()?->creatorId(), 'input_keys' => array_keys($req->all())]);
      if (!self::authorizeCompany($req, self::PERM_CREATE)) return redirect()->back();
      $valStart = microtime(true);
      $v = Validator::make($req->all(), ['name' => 'required']);
      $this->logExecutionTime($valStart, $action, 'buildValidator');
      if ($v->fails()) return redirect()->back()->with('error', $v->getMessageBag()->first());
      try {
        $type = new PerformanceType();
        $type->name = $req->input('name');
        $type[DatabaseConstants::COL_TABLE_CREATOR] = $req->user()->creatorId();
        $saveStart = microtime(true);
        $type->save();
        $this->logExecutionTime($saveStart, $action, 'savePerformanceType');
        Log::info("[{$class}::{$action}] success", ['performance_type_id' => $type->getKey(), 'name' => $type->name]);
        return redirect()->route(ViewsConstants::PFM_TP . '.index')->with('success', __('Performance Type successfully created.'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'creator_id' => $req->user()?->creatorId(), 'input_keys' => array_keys($req->all()), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function show(PerformanceType $performanceType): RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    return $this->measureProfile($action, function () use ($performanceType, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['performance_type_id' => $performanceType->getKey()]);
      try {
        $redirStart = microtime(true);
        $resp = redirect()->route(ViewsConstants::PFM_TP . '.index')->with($performanceType);
        $this->logExecutionTime($redirStart, $action, 'redirectToIndex');
        Log::info("[{$class}::{$action}] complete", ['performance_type_id' => $performanceType->getKey()]);
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'performance_type_id' => $performanceType->getKey(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        Log::error("[{$class}::{$action}] error", ['error' => $e->getMessage()]);
        throw $e;
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'performance_type_id' => $performanceType->getKey()]);
  }

  public function edit(Request $request, PerformanceType $performanceType): View|RedirectResponse
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $viewPath = ViewsConstants::PFM_TP . '.' . $action;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $performanceType, $action, $class, $viewPath) {
      Log::info("[{$class}::{$action}] start", ['performance_type_id' => $performanceType->getKey(), 'creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizeCompany($req, self::PERM_EDIT)) return redirect()->back();
      if ($performanceType[DatabaseConstants::COL_TABLE_CREATOR] !== $req->user()->creatorId()) return defaultPermissionDenial($req, new AuthorizationException(), $class . '::' . $action);
      try {
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        $renderStart = microtime(true);
        $resp = view($viewPath, compact('performanceType'));
        $this->logExecutionTime($renderStart, $action, 'renderPerformanceTypeEdit');
        Log::info("[{$class}::{$action}] complete", ['performance_type_id' => $performanceType->getKey()]);
        return $resp;
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'performance_type_id' => $performanceType->getKey(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'performance_type_id' => $performanceType->getKey()]);
  }

  public function update(Request $request, PerformanceType $performanceType): RedirectResponse|JsonResponse|null
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $performanceType, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['performance_type_id' => $performanceType->getKey(), 'creator_id' => $req->user()?->creatorId(), 'input_keys' => array_keys($req->all())]);
      if (!self::authorizeCompany($req, self::PERM_EDIT)) return redirect()->back();
      if ($performanceType[DatabaseConstants::COL_TABLE_CREATOR] !== $req->user()->creatorId()) return defaultPermissionDenial($req, new AuthorizationException(), $class . '::' . $action);
      $valStart = microtime(true);
      $v = Validator::make($req->all(), ['name' => 'required']);
      $this->logExecutionTime($valStart, $action, 'buildValidator');
      if ($v->fails()) return redirect()->back()->with('error', $v->getMessageBag()->first());
      try {
        $performanceType->name = $req->input('name');
        $saveStart = microtime(true);
        $performanceType->save();
        $this->logExecutionTime($saveStart, $action, 'savePerformanceType');
        Log::info("[{$class}::{$action}] success", ['performance_type_id' => $performanceType->getKey(), 'name' => $performanceType->name]);
        return redirect()->route(ViewsConstants::PFM_TP . '.index')->with('success', __('Performance Type successfully updated.'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'performance_type_id' => $performanceType->getKey(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'performance_type_id' => $performanceType->getKey()]);
  }

  public function destroy(Request $request, PerformanceType $performanceType): RedirectResponse|JsonResponse|null
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $performanceType, $action, $class) {
      Log::info("[{$class}::{$action}] start", ['performance_type_id' => $performanceType->getKey(), 'creator_id' => $req->user()?->creatorId()]);
      if (!self::authorizeCompany($req, self::PERM_DELETE)) return redirect()->back();
      if ($performanceType[DatabaseConstants::COL_TABLE_CREATOR] !== $req->user()->creatorId()) return defaultPermissionDenial($req, new AuthorizationException(), $class . '::' . $action);
      try {
        $delStart = microtime(true);
        $performanceType->delete();
        $this->logExecutionTime($delStart, $action, 'deletePerformanceType');
        Log::info("[{$class}::{$action}] success", ['performance_type_id' => $performanceType->getKey()]);
        return redirect()->route(ViewsConstants::PFM_TP . '.index')->with('success', __('Performance Type successfully deleted.'));
      } catch (\Throwable $e) {
        Log::debug("[{$class}::{$action}] error context", ['route' => Route::getCurrentRoute()?->getName(), 'performance_type_id' => $performanceType->getKey(), 'creator_id' => $req->user()?->creatorId(), 'message' => $e->getMessage(), 'file' => $e->getFile(), 'line' => $e->getLine()]);
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'performance_type_id' => $performanceType->getKey()]);
  }

  protected static function authorizeCompany(Request $request, string $perm): bool
  {
    if ($request->user()->{UsersConstants::COL_TP} === PermissionsConstants::CPN && $request->user()->can($perm)) return true;
    defaultPermissionDenial(
      $request,
      new AuthorizationException(),
      __CLASS__ . '::' . debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1)[0]['function']
    );
    return false;
  }
}
